control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -1047,6 +1047,13 @@ func (r *Controller) getGracefulShutdownAndUpdatePodCheck(ctx context.Context, a
 		return 0, fmt.Errorf("failed to get terminating pod %s/%s: %w", k8sNamespace, podName, err)
 	}
 
+	// In a statefulset rollout, pods can go down and a new pod will come back up with the same name but a different
+	// address. In that case, it's not the old pod that is gracefully shutting down; the old pod is gone and we
+	// should deregister that old instance from Consul.
+	if string(pod.UID) != svc.ServiceMeta[constants.MetaKeyPodUID] {
```

Comment:
I'm wondering if there is any case where the `pod.UID` or `svc.ServiceMeta[constants.MetaKeyPodUID]` are not populated and if we should account for that? 🤔

Reply:
Great call, I looked back to when we added pod.UID to service registrations and it was semi recent: https://github.com/hashicorp/consul-k8s/commit/1b6f700f17d8bc7976b6c04d92fbf2def7830e43 so older patch versions might not have it. 

I updated the check to check for either: pod uid changing, or the node name changing. So that way an upgrade from an older patch would detect node name changing even if pod uid is not yet there.